Introduction to Deep Leep Learning and Neural Networks with Keras
Week 5 Assignment
Build a Regression Model in Keras

In [1]:
# import required libraries
import pandas as pd # to read csv file into data frame  
import numpy as np  # used for working with arrays

In [2]:
# read csv file from the given link into data frame
concrete_data=pd.read_csv('https://cocl.us/concrete_data/concrete_data.csv')

In [3]:
# show first 5 rows and predictors of the dataset 
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
concrete_data.shape

(1030, 9)

Data set has 1030 samples and 8 predictors to predict the target value which is concrete strength


Concrete strength is depepndent on Cement, Blast Furnace slag, Fly Ash, Water, Superlasticizer, Coarse Aggregate, Fine Aggregate and Age of cement

Check if the data set has any missing value

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There is no null value/missing value in the given dataset

In [6]:
predictors = concrete_data[['Cement','Blast Furnace Slag','Fly Ash','Water','Superplasticizer','Coarse Aggregate','Fine Aggregate','Age']] # all columns except 'Strngth' are predictors 
target = concrete_data['Strength'] # Target is Strength

In [7]:
# check predictors
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
[row,col]=predictors.shape
print('Number of columns are: ',col)

Number of columns are:  8


In [9]:
# check target
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Normalize data by using standard method
to reduce mean of each column to zero and convert standard deviation to 1
normalized data = (data-mean of data)/standard deviation

In [10]:
# mean of predictors
m = predictors.mean()
# standard deviation of predictors
s=predictors.std()
predictors_normalized = (predictors - m) / s


In [12]:
# check normalized data
predictors_normalized.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [13]:
# import keras for builiding and fitting the model
import keras 
from keras.models import Sequential  # model type
from keras.layers import Dense  # hidden layers

In [14]:
# definition to build regression model
def regression_model():
    # sequential type
    reg_model = Sequential()
    # 10 nodes in hidden layer, number of predictor columns in input layer
    reg_model.add(Dense(10, activation='relu', input_shape=(col,)))
    # one node in output layer 
    reg_model.add(Dense(1))
    
    # compile model
    reg_model.compile(optimizer='adam', loss='mean_squared_error')
    return reg_model

Using library sklearn, rRandomly split dataframe into train and test sets 

In [15]:
from sklearn.model_selection import train_test_split

80% of the datframe is train and 20% is test
Seed of random state is fixed which uses the same split even if this code is executed multiple times  

In [23]:
X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.2, random_state= 10)

In [24]:
# build the model
model = regression_model()

In [25]:
# train model using 50 epochs
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
26/26 [==============================] - 0s 1ms/step - loss: 1496.8855
Epoch 2/50
26/26 [==============================] - 0s 2ms/step - loss: 1477.7084
Epoch 3/50
26/26 [==============================] - 0s 2ms/step - loss: 1457.7505
Epoch 4/50
26/26 [==============================] - 0s 2ms/step - loss: 1436.9232
Epoch 5/50
26/26 [==============================] - 0s 2ms/step - loss: 1414.8762
Epoch 6/50
26/26 [==============================] - 0s 2ms/step - loss: 1391.7236
Epoch 7/50
26/26 [==============================] - 0s 2ms/step - loss: 1367.3219
Epoch 8/50
26/26 [==============================] - 0s 2ms/step - loss: 1341.6827
Epoch 9/50
26/26 [==============================] - 0s 2ms/step - loss: 1314.1489
Epoch 10/50
26/26 [==============================] - 0s 2ms/step - loss: 1285.3156
Epoch 11/50
26/26 [==============================] - 0s 2ms/step - loss: 1254.8596
Epoch 12/50
26/26 [==============================] - 0s 2ms/step - loss: 1222.9479
Epoch 13/50
2

Predict concrete strength using trained model

In [26]:
y_predict = model.predict(X_test)

In [27]:
# evaluate model
# import library to calculate mean square error
from sklearn.metrics import mean_squared_error

In [29]:
mse = mean_squared_error(y_test, y_predict)
mean = np.mean(mse)
stand_dev = np.std(mse)
print('Mean of mean square error after data normalization: ', mean)
print('Standard Deviation of mean square error after data normalization: ',stand_dev )

Mean of mean square error after data normalization:  240.56533007587092
Standard Deviation of mean square error after data normalization:  0.0


In [31]:
numb_mse = 50
epochs = 50
mean_square_error = []
for i in range(0, numb_mse):
    X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.2, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    mse = model.evaluate(X_test, y_test, verbose=0)
    print("Mean Squared Error "+str(i+1)+": "+str(mse))
    y_predict = model.predict(X_test)
    mse_i = mean_squared_error(y_test, y_predict)
    mean_square_error.append(mse_i)

mean_square_error = np.array(mean_square_error)
mean = np.mean(mean_square_error)
stand_dev = np.std(mean_square_error)

print('With normalized data with number of epochs for training is: ', epochs)
print("Mean of Mean Squared error : ", mean)
print("Standard Deviation: ", stand_dev)

Mean Squared Error 1: 30.98028564453125
Mean Squared Error 2: 42.51603698730469
Mean Squared Error 3: 33.13656997680664
Mean Squared Error 4: 38.18689727783203
Mean Squared Error 5: 38.08090591430664
Mean Squared Error 6: 37.70827865600586
Mean Squared Error 7: 36.188602447509766
Mean Squared Error 8: 33.822017669677734
Mean Squared Error 9: 32.838890075683594
Mean Squared Error 10: 30.670433044433594
Mean Squared Error 11: 37.040096282958984
Mean Squared Error 12: 32.930912017822266
Mean Squared Error 13: 36.56557846069336
Mean Squared Error 14: 45.1212158203125
Mean Squared Error 15: 37.25030517578125
Mean Squared Error 16: 30.931026458740234
Mean Squared Error 17: 35.09627151489258
Mean Squared Error 18: 36.912925720214844
Mean Squared Error 19: 33.64012908935547
Mean Squared Error 20: 40.337093353271484
Mean Squared Error 21: 34.02544403076172
Mean Squared Error 22: 37.434749603271484
Mean Squared Error 23: 30.181385040283203
Mean Squared Error 24: 37.78107833862305
Mean Squared Er

1. Mean of Mean Squared error without normalization of data is 50.22 and with normalization is 35.28
2. It can be concluded that normaization of data reduces mean squared error
